GooglePlayInstant/LoadingScreen/LoadingScreenScript.cs

```diff
@@ -56,27 +59,39 @@ private IEnumerator Start()
             }
         }
 
-        //TODO: Update function for unity 5.6 functionality
         private IEnumerator GetAssetBundle(string assetBundleUrl)
         {
-#if UNITY_2018_2_OR_NEWER
-            var www = UnityWebRequestAssetBundle.GetAssetBundle(assetBundleUrl);
+#if UNITY_2018_1_OR_NEWER
+            var webRequest = UnityWebRequestAssetBundle.GetAssetBundle(assetBundleUrl);
+            var assetbundleDownloadOperation = webRequest.SendWebRequest();
 #else
-            var www = UnityWebRequest.GetAssetBundle(assetBundleUrl);
+            var webRequest = UnityWebRequest.GetAssetBundle(assetBundleUrl);
+            var assetbundleDownloadOperation = webRequest.Send();
 #endif
-            var assetbundleDownloadOperation = www.SendWebRequest();
-
             yield return StartCoroutine(LoadingBar.UpdateLoadingBar(assetbundleDownloadOperation,
                 LoadingBar.AssetBundleDownloadMaxWidthPercentage));
 
-            // TODO: implement retry logic
-            if (www.isNetworkError || www.isHttpError)
+#if UNITY_2017_1_OR_NEWER
+            if (webRequest.isHttpError || webRequest.isNetworkError)
+#else
+            if (webRequest.isError)
+#endif
             {
-                Debug.LogErrorFormat("Error downloading asset bundle: {0}", www.error);
+                if (_assetBundleRetrievalAttemptCount < MaxAttemptCount)
+                {
+                    _assetBundleRetrievalAttemptCount++;
+                    Debug.LogFormat("Attempt #{0} at downloading AssetBundle...", _assetBundleRetrievalAttemptCount);
+                    yield return new WaitForSeconds(2);
+                    yield return GetAssetBundle(assetBundleUrl);
```

Comment:
My C#-foo is pretty shoddy, but shouldn't this be

```
for (var bar in GetAssetBundle(...)) {
  yield bar;
}
```

IIRC `yield return GetAssetBundle(...)` would try to return an `IEnumerator<IEnumerator<?>>`

Reply:
Here I'm following documentation from Unity that describes how to start coroutines, so that pieces of asynchronous code can be executed synchronously and in separate Update() frames: https://docs.unity3d.com/ScriptReference/MonoBehaviour.StartCoroutine.html
https://docs.unity3d.com/Manual/Coroutines.html

This line was necessary (yield return) because ignoring the return value of a method that returns an IEnumerator (a coroutine) wouldn't let the coroutine would not continue if the return value was ignored.

Overall though, I'm not very familiar with the syntax in the block above, and haven't seen it in other pieces of Unity C# code (but it does look like UnityScript/JS).

Reply:
nvm, I followed through the explanation on https://stackoverflow.com/questions/12932306/how-does-startcoroutine-yield-return-pattern-really-work-in-unity, in this case, shouldn't this be

```
yield return StartCoroutine(GetAssetBundle(assetBundlerUrl));
```

otherwise, Unity's event loop will get the sequence of types:

`Coroutine` // it will happily run this on the next update
`WaitForSeconds` // it will happily wait on this for 2 seconds after that update
`IEnumerable` // unity won't know what to do with this and will just drop it
terminated

on the other hand, `yield return StartCoroutine(IEnumerableObject)` will wrap the `GetAssetBundle(...)` `IEnumerable` within a Unity `Coroutine` type, and its eventloop will understand that it should execute the coroutine after the next update.

If we did indeed do

```
foreach (var x in GetAssetBundle(...)) {
  yield return x;
}
```

we would essentially do the retry logic all within the same Update frame, which is definitely not desired.

In anycase, just wrap the GetAssetBundle within `StartCoroutine(GetAssetBundle(...))` should be correct.

Reply:
I was looking into this before, and I was surprised to learn that:

```
yield return StartCoroutine(GetAssetBundle(assetBundlerUrl));
```

and

```
yield return GetAssetBundle(assetBundlerUrl);
```

are equivocally the same now in Unity after an update to 5.4:
https://answers.unity.com/questions/724244/yield-return-m-vs-yield-return-startcoroutine.html
https://forum.unity.com/threads/difference-between-yield-return-ienumerator-and-yield-return-startcoroutine.432571/

Do you think yield return StartCoroutine(...) is more clear though? I could see how this could become frustrating later.

Reply:
Woops, that's a good catch, let's actually just drop StartCoroutine altogether if IEnumerable can be automatically lifted to a coroutine, it's also clearer where the control flow is being redirected to.

Reply:
Sounds good! Resolved!